docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,838 @@
+
+        /* Call enclave function. */
+        if ((_result = oe_call_enclave_function(
+                 enclave,
+                 &global_id,
+                 __test_ecall_info_table[test_fcn_id_enc_foo].name,
+                 _input_buffer,
+                 _input_buffer_size,
+                 _output_buffer,
+                 _output_buffer_size,
+                 &_output_bytes_written)) != OE_OK)
+            goto done;
+
+        /* Setup output arg struct pointer. */
+        _pargs_out = (foo_args_t*)_output_buffer;
+        OE_ADD_SIZE(_output_buffer_offset, sizeof(*_pargs_out));
+    
+        /* Check if the call succeeded. */
+        if ((_result = _pargs_out->_result) != OE_OK)
+            goto done;
+
+    -   /* When there are out-only, deep-copyable parameters, _output_bytes_written
+    -    * must be large than or equal to _output_buffer_size. */
+    -   if (_output_bytes_written < _output_buffer_size)
+    +   /* Currently exactly _output_buffer_size bytes must be written. */
+    +   if (_output_bytes_written != _output_buffer_size)
+        {
+            _result = OE_FAILURE;
+            goto done;
+        }
+    +   /* Calculate the size of the buffer, which holds the deep-copied content. */
+    +   _deepcopy_out_buffer_size = _output_bytes_written - _output_buffer_size;
+    +    if (_deepcopy_out_buffer_size)
+    +   {
+    +       _deepcopy_out_buffer = _pargs_out->_host_buffer;
+    +       if (!_deepcopy_out_buffer)
+    +       {
+    +           _result = OE_FAILURE;
+    +           goto done;
+    +       }
+    +   }
+
+        /* Unmarshal return value and out, in-out parameters. */
+        /* No return value. */
+
+    +   OE_READ_OUT_PARAM(ptr, (size_t)(((size_t)1 * sizeof(NestedBlob))));
+    +   if (ptr)
+    +       OE_SET_DEEPCOPY_OUT_PARAM(ptr->blob_array, ((size_t)ptr->num * sizeof(Blob)), Blob*);
+    +   for (size_t _i_2 = 0; _i_2 < ptr->num; _i_2++)
+    +   {
+    +       if (ptr && ptr->blob_array)
+    +           OE_SET_DEEPCOPY_OUT_PARAM(ptr->blob_array[_i_2].buf, ptr->blob_array[_i_2].len, char*);
+    +   }
+
+    +   if (_deepcopy_out_buffer_offset != _deepcopy_out_buffer_size)
+    +   {
+    +       _result = OE_FAILURE;
+    +       goto done;
+    +   }
+
+        _result = OE_OK;
+
+    done:
+        if (_buffer)
+            free(_buffer);
+
+    +   if (_deepcopy_out_buffer)
+    +       free(_deepcopy_out_buffer);
+
+        return _result;
+    }
+    ```
+
+  - Enclave
+
+    Upon the dispatching of the ECALL, the oeedger8r-generated
+    `ecall_foo` is invoked as usual. After the invocation of
+    the user-defined `foo` function, the enclave calculates the size
+    of content set by `foo`. Next, the enclave allocates an in-enclave
+    buffer, `_deepcopy_out_buffer`, and serializes the content into the buffer using the
+    `OE_WRITE_DEEPCOPY_OUT_PARAM` macro.
+    After serializing, the enclave allocates a host buffer (held by the `pargs_out->_host_buffer`
```

Comment:
This does not have to be done by the generated code. 
The generated code just needs to do 
```c
    args->_host_buffer = _deepcopy_out_buffer;
    args->_host_buffer_size = _deepcopy_out_buffer_size;
```

The ECALL dispatching infrastructure can determine the best way to marshal _host_buffer over to the host, according to the platform (SGX/OPTEE).

Reply:
Updated.